### PR TITLE
G1Affine to/from bytes

### PR DESCRIPTION
**Pull request by vlopes11**
Implement unsafe creation of to/from bytes optimized for performance

Implements #44 and #46

### CODE REVIEW

File: src/g1.rs [hunk 1]
```diff
@@ -257,9 +257,57 @@ impl G1Affine {
         }
     }
 
+    /// Raw bytes representation
+    ///
+    /// The intended usage of this function is for trusted sets of data where performance is
+    /// critical. This way, the `infinity` internal attribute will not be stored and `x` and
+    /// `y` will be stored without any check.
+    ///
+    /// For secure serialization, check `to_compressed` and `to_uncompressed`
```
Comment (CPerezz) [inline src/g1.rs:266]: Not sure `to_compressed` and `to_uncompressed` exist anymore. Also, `to_bytes` and `from_bytes` should be the referenced fn's.

File: src/g1.rs [hunk 2]
```diff
@@ -282,101 +330,12 @@ impl G1Affine {
...
-                    (!sort_flag_set),
-                )
-            })
-        })
-    }
-
     /// Attempts to deserialize a compressed element. See [`notes::serialization`](crate::notes::serialization)
     /// for details about how group elements are serialized.
-    pub fn from_compressed(bytes: &[u8; 48]) -> CtOption<Self> {
+    pub fn from_bytes(bytes: &[u8; 48]) -> CtOption<Self> {
```
Comment (CPerezz) [inline src/g1.rs:335]: Do we want this to return a `CtOption` still? I thought we were going to return the correct type directly.

**Review by CPerezz — CHANGES_REQUESTED** [review]
Couple of questions:
1) Weren't we going to introduce `dusk-bytes` here?
2) The tests are not passing when `serde` flag is active. 
The method `to_compressed` doesn't exist anymore. So the serde impl needs to call the `from/to_bytes`. 

Also, we should state that the serde impls work with the "safe" byte reprs. There are no plans to support unchecked bytes on serde.

File: src/g1.rs (hunk 2 above)
Comment (vlopes11) [reply src/g1.rs:335]: I don't want, but I believe this `CtOption` problem should be handled by a different issue

File: src/g1.rs [hunk 3]
```diff
@@ -54,6 +55,99 @@ impl From<G1Projective> for G1Affine {
     }
 }
 
+impl Serializable<48> for G1Affine {
```
Comment (CPerezz) [inline src/g1.rs:58]: Instead of `48` raw, why not using `G1_COMPRESSED_SIZE`?

File: src/g1.rs [hunk 4]
```diff
@@ -54,6 +55,99 @@ impl From<G1Projective> for G1Affine {
...
+    type Error = BytesError;
+
+    /// Serializes this element into compressed form. See [`notes::serialization`](crate::notes::serialization)
+    /// for details about how group elements are serialized.
+    fn to_bytes(&self) -> [u8; 48] {
```
Comment (CPerezz) [inline src/g1.rs:63]: Same, Why not use and export a const with the size?

File: src/g1.rs [hunk 5]
```diff
@@ -54,6 +55,99 @@ impl From<G1Projective> for G1Affine {
...
+            &(1u8 << 5),
+            (!self.infinity) & self.y.lexicographically_largest(),
+        );
+
+        res
+    }
+
+    /// Attempts to deserialize a compressed element. See [`notes::serialization`](crate::notes::serialization)
+    /// for details about how group elements are serialized.
+    fn from_bytes(buf: &[u8; 48]) -> Result<Self, Self::Error> {
```
Comment (CPerezz) [inline src/g1.rs:88]: Ditto

File: src/g1.rs [hunk 6]
```diff
@@ -257,177 +347,52 @@ impl G1Affine {
...
-        // We already know the point is on the curve because this is established
-        // by the y-coordinate recovery procedure in from_compressed_unchecked().
+    /// Raw bytes representation
+    ///
+    /// The intended usage of this function is for trusted sets of data where performance is
+    /// critical. This way, the `infinity` internal attribute will not be stored and `x` and
+    /// `y` will be stored without any check.
+    ///
+    /// For secure serialization, check `to_bytes`
+    pub fn to_raw_bytes(&self) -> [u8; 96] {
```
Comment (CPerezz) [inline src/g1.rs:357]: Use and export a const with the serialized size.

File: src/g2.rs [hunk 7]
```diff
@@ -88,6 +90,53 @@ impl PartialEq for G2Affine {
     }
 }
 
+impl Serializable<96> for G2Affine {
```
Comment (CPerezz) [inline src/g2.rs:93]: Use constants and export them such as `G2_COMPRESSED_SIZE` which is unused.

File: src/g2.rs [hunk 8]
```diff
@@ -88,6 +90,53 @@ impl PartialEq for G2Affine {
...
+    type Error = BytesError;
+
+    /// Serializes this element into compressed form. See [`notes::serialization`](crate::notes::serialization)
+    /// for details about how group elements are serialized.
+    fn to_bytes(&self) -> [u8; 96] {
```
Comment (CPerezz) [inline src/g2.rs:98]: ditto

File: src/g2.rs [hunk 9]
```diff
@@ -88,6 +90,53 @@ impl PartialEq for G2Affine {
...
+            &(1u8 << 5),
+            (!self.infinity) & self.y.lexicographically_largest(),
+        );
+
+        res
+    }
+
+    /// Attempts to deserialize a compressed element. See [`notes::serialization`](crate::notes::serialization)
+    /// for details about how group elements are serialized.
+    fn from_bytes(buf: &[u8; 96]) -> Result<Self, Self::Error> {
```
Comment (CPerezz) [inline src/g2.rs:128]: ditto

File: src/g2.rs [hunk 10]
```diff
@@ -289,34 +334,65 @@ impl G2Affine {
...
-        // Is this point at infinity? If so, set the second-most significant bit.
-        res[0] |= u8::conditional_select(&0u8, &(1u8 << 6), self.infinity);
+    /// Raw bytes representation
+    ///
+    /// The intended usage of this function is for trusted sets of data where performance is
+    /// critical. This way, the `infinity` internal attribute will not be stored and `x` and
+    /// `y` will be stored without any check.
+    ///
+    /// For secure serialization, check `to_bytes`
+    pub fn to_raw_bytes(&self) -> [u8; 192] {
```
Comment (CPerezz) [inline src/g2.rs:344]: ditto

File: src/g2.rs [hunk 11]
```diff
@@ -88,6 +90,53 @@ impl PartialEq for G2Affine {
...
+        res
+    }
+
+    /// Attempts to deserialize a compressed element. See [`notes::serialization`](crate::notes::serialization)
+    /// for details about how group elements are serialized.
+    fn from_bytes(buf: &[u8; 96]) -> Result<Self, Self::Error> {
+        // We already know the point is on the curve because this is established
+        // by the y-coordinate recovery procedure in from_compressed_unchecked().
+
+        let x: Option<Self> = Self::from_compressed_unchecked(buf).into();
```
Comment (CPerezz) [inline src/g2.rs:132]: `from_compressed_unchecked` should also dissappear IMO. I would add everything in `from_bytes` directly.

File: Cargo.toml [hunk 12]
```diff
@@ -18,17 +18,21 @@ exclude = [".github/workflows/ci.yml",
...
 rayon = "1.3.0"
 serde = {version = "1.0.106", optional = true}
-rand_core = "0.6"
+rand_core = "0.5"
 canonical = {version = "0.4", optional = true}
 canonical_derive = {version = "0.4", optional = true}
+dusk-bytes = {git="https://github.com/dusk-network/dusk-bytes", tag="v0.1.0"}
+
+# Required to fix the different version canonical_derive and dusk-bytes use for syn
+syn = "1.0.58"
```
Comment (CPerezz) [inline Cargo.toml:27]: We should solve this. In dusk-bytes if possible.

**Review by CPerezz — COMMENTED** [review]
A couple of requests/suggestions.

**Review by ZER0 — CHANGES_REQUESTED** [review]
Overall looks good to me, but I will adjust few things noted in the comments.

Also, since it seems you have `from_slice_unchecked`, I would implement the trait from `dusk-bytes` to have `from_slice` too for the type that implements `to_bytes` and `from_bytes`, in order to be consistent.

And keep in mind that if you're using `dusk-bytes` you can implement, even just for test, the trait `ParseHexStr` for free, so you can have:

```rust
assert!(Scalar::from_hex_str("000fffffffffe5bfeff2a4bd535d8a198d83933487d9d2953a7ed73").is_err())
```

File: src/g1.rs (hunk 6 above)
Comment (vlopes11) [reply src/g1.rs]: Done with G1Affin::RAW_SIZE

File: src/g2.rs (hunk 8 above)
Comment (vlopes11) [reply src/g2.rs]: Done with G1Affine::RAW_SIZE

File: src/g2.rs (hunk 9 above)
Comment (vlopes11) [reply src/g2.rs]: Same as above

File: src/g2.rs (hunk 10 above)
Comment (vlopes11) [reply src/g2.rs]: Done with G2Affine::RAW_SIZE

File: src/g2.rs (hunk 11 above)
Comment (vlopes11) [reply src/g2.rs]: Agreed, done

**Review by julesdesmit — APPROVED** [review]
LGTM

File: Cargo.toml (hunk 12 above)
Comment (vlopes11) [reply Cargo.toml]: Fixed

File: src/fp.rs [hunk 13]
```diff
@@ -312,6 +312,11 @@ impl Fp {
         Fp(v)
     }
 
+    /// Internal representation of `Fp`
+    pub const fn internal_repr(&self) -> &[u64; 6] {
+        &self.0
+    }
```
Comment (ZER0) [inline src/fp.rs:318]: Not to solve in this PR, but this method always bothered me; we should go for a more idiomatic rust naming convention. What do you think @vlopes11 ?

File: src/scalar.rs [hunk 14]
```diff
@@ -905,15 +906,15 @@ fn test_inv() {
 fn test_debug() {
     assert_eq!(
         format!("{:?}", Scalar::zero()),
-        "0x0000000000000000000000000000000000000000000000000000000000000000"
+        "0000000000000000000000000000000000000000000000000000000000000000"
```
Comment (ZER0) [inline src/scalar.rs:909]: Notice that you can keep the `0x` form just adding the alternate symbol to the format, so where `{:?}` produces "0000…" (etc); `{:#?}` produces `0x0000`.

**Review by ZER0 — APPROVED** [review]
LGTM

File: src/fp.rs (hunk 13 above)
Comment (vlopes11) [reply src/fp.rs:318]: As discussed, will create an issue for that

File: src/scalar.rs (hunk 14 above)
Comment (vlopes11) [reply src/scalar.rs:909]: Both tested and working fine

**Pull request merged**